### PR TITLE
Mm 38361 pre written messages

### PR DESCRIPTION
**Pull request by neallred**
#### Summary
Add prewritten messages A/B test. A/B test variations are controlled at a server level in Split and require having someone with access changing the `PrewrittenMessages` value in split. You can ask me to change the value to see the different variations.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-38361

#### Related Pull Requests
- server changes: https://github.com/mattermost/mattermost-server/pull/18372

#### Screenshots

control group:
prewritten-none

prewritten messages in tip:
prewritten-tour-point

prewritten messages by create post box:
prewritten-around-input

prewritten messages to self:
prewritten-self

prewritten messages to group:
prewritten-group

#### Release Note

```release-note
Added A/B test for providing user with prewritten messages.
```

### CODE REVIEW

**Comment by annelieseklein** [comment]
@neallred looks great! For all instances can we please change the :hand: to :wave: ? For the "Oh hello" can we change "🙌" to " ✌️ "?  Also for test C, the highlight box should be at 8% opacity of the button color

**Comment by neallred** [comment]
Having issues getting the spinwick to pull in the feature flag from split. Will post here when its resolved.

File: components/create_post/prewritten_chips.tsx
Comment (pablovelezvidal) [inline components/create_post/prewritten_chips.tsx:107]: nit: if we are using styled components, we should try to avoid inline styles

File: components/create_post/create_post_tip.tsx
Comment (pablovelezvidal) [inline components/create_post/create_post_tip.tsx:11]: group mattermost-redux imports together

File: components/create_post/create_post.tsx
Comment (pablovelezvidal) [inline components/create_post/create_post.tsx:1361]: is this div used to layout the internal structure, or just to accomplish the react rule of only render one tag? if so, maybe you could use a React.Fragment `<></>` instead.

File: components/create_post/create_post.tsx
Comment (pablovelezvidal) [inline components/create_post/create_post.tsx:618]: nit: I think it is a good practice to make code self documented and put this kind of long conditions in a constant value to make code self explanatory and easy to read, for example, in this case, it could be something like `const shouldShowPrewrittenMsgs = this.props.tutoriaStep === Contants.Tuto....`

**Review by pablovelezvidal — APPROVED** [review]
Great Work @neallred . Just some minor comments!

**Comment by neallred** [comment]
Got the spinwick updated with the right server code! You can see the pulsating dot version of the test now.

**Comment by esethna** [comment]
Thanks @neallred,

For this version

image
1. Can we update the copy?

Select or type your first message and select **Enter** to send it. 

Use the **Attachments** and **Emoji** buttons to add to your messages.

@annelieseklein feel free to ping me to review copy on future designs ahead of implementation. 

2. @annelieseklein @neallred it looks like the buttons are missing hover states? 

3. Did we add telemetry to each of the buttons, including the "Got it" button and dismissing the modal without using a button (ie esc or click away)? 

4. The designs state that the tips "reengage" after 48 hours, what does that mean?

**Comment by neallred** [comment]
> 1. ...copy?

Sure!

> 2. missing hover states

They were incorrectly coded on my part.

> 3. ... telemetry to each...

The got it button has it, but I still need to add to each of the others.

> 4.  designs state that the tips "reengage" after 48 hours ...

The ticket should probably be updated. It was my understanding this was a potential later refinement that we don't want to implement now.

**Comment by AGMETEOR** [comment]
@neallred I want to approve but I feel like I should wait until all the PM feedback is completed since there would be code changes.

**Review by esethna — APPROVED** [review]
Approving pending copy updates

**Comment by neallred** [comment]
@pablovelezvidal re-requested review because there's been some meaningful changes to tracking users' dismissal of tutorial tip.

File: components/tutorial/tutorial_tip/tutorial_tip.tsx
Comment (neallred) [inline components/tutorial/tutorial_tip/tutorial_tip.tsx:88]: Per conversation with @esethna we want to track when the user hides the modal. I was hoping this function could be simpler, but for cases like prewritten messages where we need to generate a click to hide the overlay, this function gets called twice, once for the original element click, and once for the create post textbox being clicked/focused. Also, this function gets called when the user clicks "got it" or opts out of the tips. So I had to do some event sniffing to tell if the user was really dismissing the overlay.

**Review by annelieseklein — APPROVED** [review]

**Comment by annelieseklein** [comment]
> The ticket should probably be updated. It was my understanding this was a potential later refinement that we don't want to implement now.

^Yes this was for future state -- I'll update the ticket to remove it.

**Review by AGMETEOR — APPROVED** [review]
Awesome job @neallred 🎉 LGTM

File: components/create_post/chip.tsx
Comment (neallred) [reply components/create_post/chip.tsx]: I web searched it, and I guess its `free-form`: https://www.merriam-webster.com/dictionary/free-form

**Review by pablovelezvidal — APPROVED** [review]
Great work @neallred ! LGTM :shipit:

**Review by stevemudie — APPROVED** [review]
All of the various variations checked and looking as expected.

Great work @neallred! You get to stay another day!

**Pull request merged**
